### DISCUSSION TITLE
OME soft locking after some time

### DISCUSSION THREAD

**Discussion by naanlizard** [OvenMediaLabs/OvenMediaEngine#1160] [Help]
I'm not 100% sure this isn't our stream forwarding method from nginx-rtmp to OME, but I am pretty confident it is OME because when it happens, if I restart our forwarder, it doesn't fix it, but restarting OME does fix it.

Essentially, this screenshot shows the problem, but I'm not sure if it is cause or effect. CPU usage drops to almost nothing, nobody can play streams, no crash dump is created.

<img width="709" alt="Screenshot 2023-03-30 at 8 31 23 AM" src="https://user-images.githubusercontent.com/7061721/228762409-ee25eb88-bdd6-444d-9b8e-9397dbcf960b.png">

CPU usage doesn't match inside/outside the docker either. When I run `docker exec omedocker ps aux` I get 223% CPU, but `docker stats` reports .33%

Those examples were from a time this happened on the 26th of March. Today it happened again just now, hence the discussion.

Here is the relevant info

[Server.xml](https://gist.github.com/naanlizard/8b5745c02dc82cb2f21ae6b1e3512ef0)

[Dockerfile](https://gist.github.com/naanlizard/821de93ba7653aad981f7d83080aa6cf)

[Logger.xml](https://gist.github.com/naanlizard/d16ded99ed1d2fd2769428d211f88cce)

[EdgeConfig.xml](https://gist.github.com/naanlizard/b61020b7397202ecf8dc1addd596429d)

[Logs from today](https://gist.github.com/naanlizard/cb19bb1c4794b0335d033f06f2862543) - crash happened around 05:3X in the timestamps there. - 05:33:45 on my grafana instance, but I'm unsure how accurate that is

**Comment by getroot** [comment]
Which OME version are you using?

**Reply by naanlizard** [reply]
From the dockerfile - FROM airensoft/ovenmediaengine:0.15.7 :)

**Reply by getroot** [reply]
Do you have viewers playing in HLS, DASH, LLDASH from your OME?

**Reply by naanlizard** [reply]
Viewers can play Legacy HLS (default), WebRTC, and LLHLS as well as an audio only WebRTC we provide, at any time streams are live.

**Reply by getroot** [reply]
Yes... Most of these problems are caused by a bug on the publisher side that is not responding to player requests and is holding them (infinite loop or mutex deadlock). In this case, a dead lock occurs on the PhysicalPort, which is a module in charge of sockets, and new connections cannot be accepted.

If Legacy HLS Publisher has such problems, it's not easy to troubleshoot. (Of course, I don't know where the problem is yet.)

LLHLS Publisher can also service HLS in legacy mode as shown below.
https://domain/app/stream/llhls.m3u8?_HLS_legacy=yes

By the way, there is no log of 05:3x time in the log file. Didn't you upload another file?

**Reply by getroot** [reply]
And there are two files, Server.xml and EdgeConfig.xml. Is the deadlocked server Origin or Edge?

**Reply by naanlizard** [reply]
We run one server, just Origin. I included EdgeConfig for completeness, apologies for the confusion.

We do not use legacy HLS from OME, apologies I misremembered. Legacy HLS is served from nginx-rtmp still, only llhls and webrtc are from OME

Re: log files, I meant the X as a wildcard - I'm not sure exactly when it locked. But you can ctrl-f to search the file for `05:30:05.838` and that is approximately the correct time

**Reply by getroot** [reply]
Oh, I see.

If there is a problem with a specific Publisher, the port used by that Publisher will be blocked. If any Publisher is using a different port, it will continue to work. Is it possible to check if all of your WebRTC (3333, 3334 ports) and LLHLS (799, 800 ports) are blocked in a deadlock situation? This will be an important clue for us to analyze the problem.

We fixed an issue yesterday where the thumbnail publisher could get deadlocked under certain conditions (regression issue), but looking at your Server.xml it seems that you are not using the thumbnail publisher.

**Reply by naanlizard** [reply]
I'm unsure what you mean by ports blocked - I can't play either webrtc or llhls when this happens, we only use the SSL ports for this (800 and 3334)

What do you want me to test when it happens?

**Reply by getroot** [reply]
I would like to know if both port 800 and port 3334 are blocked or only one is blocked in a deadlock situation.

**Reply by naanlizard** [reply]
Neither serves streams when this happens, I'm not sure if that answers your question. Later I can post web console errors for each if that helps

**Reply by getroot** [reply]
okay. Since usually only one port 800 or one port 3334 is blocked in this case, we can figure out which publisher is causing the problem. For example, if a WebRTC publisher has this kind of problem, a LLHLS publisher can work just fine. If both don't work, we can guess that it's not the publisher's problem, but the lower module's problem.

**Reply by getroot** [reply]
Looking at the log you posted, since around 05:3x, I can't find any new LLHLS playback requests, but WebRTC is still receiving playback requests. Inference based on this, maybe there is a deadlock problem on the LLHLS side.
I have a few more questions to make finding the problem easier. (sorry to bother you)

How long after turning on the server did this problem reproduce?

Are there any versions that do not experience this problem? If so, which of those versions is the most recent?

**Reply by naanlizard** [reply]
Please, ask anything. 

I'll review our logs but this has been happening for a while on at least two versions,this happened after only four days uptime. Usually it is not so frequent

**Reply by getroot** [reply]
I got one clue from your logs.
This issue will be resolved soon.

**Reply by naanlizard** [reply]
This is very intimidating haha

I'm very curious what the problem is, I'm looking forward to the commit. Thank you for the hard work!

**Reply by naanlizard** [reply]
Please let me know what docker hash I should pull (or if I should wait for 0.15.8) and I'll test ASAP

**Reply by getroot** [reply]
It's still just a clue, so I'm not sure. I'll get back to you when the analysis is complete.

**Comment by getroot** [comment]
https://github.com/AirenSoft/OvenMediaEngine/commit/02876e43c137b75b57aee81cfa6be6b011def27c

I can't confirm this yet, but it fixed what I suspect was the cause of your problem. Is it difficult to test the dev version?

The cause of the problem is freezing when attempting to delete an LLHLS stream with a very low probability. I'll look into this further.

**Reply by naanlizard** [reply]
I'm now running the latest commit (DVR logging change) and will report back if it soft locks again.

We have a lot of streams start/stop all the time (is an input stream stopping the same as deleting?), so I can imagine we would trigger it. Especially with a 16c/32t CPU (if my amateur guess is that more cores = more chances to break)

**Reply by getroot** [reply]
yes, deleting stream is same as stopping in ome. 
Thanks for testing the dev version.  Hope this solves your problem!

**Reply by naanlizard** [reply]
This hasn't happened since I posted I don't think, we can probably consider it solved!
